lib/events/stream.go

```diff
@@ -0,0 +1,694 @@
+
+	// stream is the GRPC stream off of which events are consumed.
+	stream proto.AuthService_StreamSessionRecordingServer
+
+	// serverID is the identity of the server extracted from the x509 certificate.
+	serverID string
+
+	// sessionID is the unique ID of the session.
+	sessionID string
+
+	// writer is a io.Pipe writer over which writes to the tarball are done.
+	writer io.WriteCloser
+
+	// reader is a io.Pipe reader over which reads from the tarball are read
+	// by the uploader.
+	reader io.ReadCloser
+
+	// tarWriter is used to create the archive itself.
+	tarWriter *tar.Writer
+
+	// zipBuffer
+	zipBuffer *bytes.Buffer
+
+	// zipWriter is used to create the zip files within the archive.
+	zipWriter *gzip.Writer
+
+	// uploadContext is used to cancel the tarball upload.
+	uploadContext context.Context
+
+	// uploadCancel is used to cancel the tarball upload.
+	uploadCancel context.CancelFunc
+
+	// waitCh is used to unblock when the upload completes and return
+	// the error (or nil).
+	waitCh chan error
+}
+
+// NewStreamManger is used to manage common stream resources like a pool of
+// buffers and a semaphore.
+func NewStreamManger(ctx context.Context) *StreamManager {
+	// If no context is passed in (like in tests) then set a background context.
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return &StreamManager{
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "stream",
+		}),
+		pool: sync.Pool{
+			New: func() interface{} {
+				return new(bytes.Buffer)
+			},
+		},
+		semaphore:    make(chan struct{}, defaults.ConcurrentStreams),
+		closeContext: ctx,
+	}
+}
+
+// ProcessStream reads events off a GRPC stream and processes them.
+func (s *StreamManager) ProcessStream(serverID string, uploader SessionUploader, stream proto.AuthService_StreamSessionRecordingServer) (*Stream, error) {
+	// Wrap the parent process shutdown context and create an upload context that is
+	// used by the uploader to cancel a upload if an error occurs while creating
+	// the archive.
+	ctx, cancel := context.WithCancel(s.closeContext)
+
+	st := &Stream{
+		manager:       s,
+		stream:        stream,
+		chunkType:     typeInit,
+		uploader:      uploader,
+		serverID:      serverID,
+		uploadContext: ctx,
+		uploadCancel:  cancel,
+		waitCh:        make(chan error),
+	}
+
+	// Start reading off the stream and processing events.
+	go st.start()
+
+	return st, nil
+}
+
+// takeSemaphore will acquire the semaphore.
+func (s *StreamManager) takeSemaphore() error {
+	select {
+	case s.semaphore <- struct{}{}:
+		return nil
+	case <-s.closeContext.Done():
+		return errContext
+	}
+}
+
+// releaseSemaphore will release the semaphore.
+func (s *StreamManager) releaseSemaphore() error {
+	select {
+	case <-s.semaphore:
+		return nil
+	case <-s.closeContext.Done():
+		return errContext
+	}
+}
+
+// Wait will block until the stream has been processed and a response written
+// to the client.
+func (s *Stream) Wait() error {
+	return <-s.waitCh
+}
```

Comment:
Additional calls to `Wait` will block forever.  This is ok, but potentially confusing since `Wait` is conventionally a concurrency-safe method.  I think `Wait` should either explicitly warn that it is only safe to be called once, or it should be tweaked to be callable from multiple goroutines (e.g. with a `done`-style signal channel and a mutex-protected `err` field).